### PR TITLE
2017wk38

### PR DESCRIPTION
**Pull request by kerolasa**
Here is latest set of improvements. Each commit should have explanation why they are needed, if not that needs to be fixed before merge.

### CODE REVIEW

**Comment by nmav** [comment]
Could you check the travis failure? As far as I understand you are eliminating compiler warnings. Wouldn't it make sense to enhance the travis CI with a -Wall and -Werror build? Possibly a static analyzer too, to prevent the introduction of such issues in the future?

**Comment by kerolasa** [comment]
That travis failure does not tell too much why it happen. Let me setup my own debugging travis and sort that out (not today, but lets say within a week).

I will leave adding -Wall and -Werror to you, as they are policy decisions and I think it's maintainer who needs to choose the level of strictness.

**Comment by kerolasa** [comment]
Oh my. I must be happy that there is automatic tests, the issue was correctly reported and as silly as bugs go. Call to atexit(function) always returned none-zero return value. That is now fixed to fail only when appropriate.

File: ipcalc.c [hunk 1]
```diff
@@ -968,9 +968,10 @@ int get_ipv6_info(const char *ipStr, int prefix, ip_info_st * info,
 	return 0;
 }
 
-static int randomize(void *ptr, unsigned size)
+static int randomize(void *ptr, ssize_t size)
```
Comment (nmav) [inline ipcalc.c]: Why would you allow providing a negative value as size?

File: ipcalc.c [hunk 2]
```diff
@@ -1598,5 +1598,19 @@ int main(int argc, char **argv)
...
+		free(info.hostmin);
+	free(info.ip);
+	free(info.netmask);
+	if (info.hostmax != info.network)
+		free(info.network);
+	free(info.reverse_dns);
+	free(splitStr);
+
 	return 0;
 }
```
Comment (nmav) [inline ipcalc.c]: r+

File: ipcalc.c [hunk 3]
```diff
@@ -118,8 +118,6 @@ extern char __attribute__((warn_unused_result)) *safe_strdup(const char *str)
 */
 uint32_t prefix2mask(int prefix)
 {
-	struct in_addr mask;
-	memset(&mask, 0, sizeof(mask));
```
Comment (nmav) [inline ipcalc.c]: r+

File: ipcalc-tests [hunk 4]
```diff
@@ -34,24 +34,24 @@ fail() {
...
+	if [ $rc -eq 0 ]; then fail "$output"; else ok; fi
 }
 
 TestOutput() {
 	echo -n "Checking $1... "
 	output=$(sh -c "$1" 2>&1)
 	rc=$?
-	[ "$output" = "$2" ] && ok || fail $output
+	if [ "$output" = "$2" ]; then ok; else fail "$output"; fi
 }
```
Comment (nmav) [inline ipcalc-tests]: r+

File: ipcalc.1 [hunk 5]
```diff
@@ -94,57 +94,65 @@ The variable exposed is PREFIX.
...
+.br
+.MT pbrown@redhat.com
+Preston Brown
+.ME
+.br
+.MT dcantrell@redhat.com
+David Cantrell
+.ME
 .SH "REPORTING BUGS"
-Report bugs at https://github.com/nmav/ipcalc/issues
```
Comment (nmav) [inline ipcalc.1]: r+

File: ipcalc.c [hunk 6]
```diff
@@ -599,7 +599,7 @@ char *ipv6_prefix_to_hosts(char *hosts, unsigned hosts_size, unsigned prefix)
 #define FLAGS_TO_IGNORE_MASK (~FLAGS_TO_IGNORE)
 
 static
-int get_ipv4_info(const char *ipStr, int prefix, ip_info_st * info,
```
Comment (nmav) [inline ipcalc.c]: As it is never modified, I think having the ipStr marked as const is the right thing here.

File: ipcalc.c [hunk 7]
```diff
@@ -1284,18 +1272,39 @@ int main(int argc, char **argv)
...
 		if (optind < argc)
 			chptr = argv[optind++];
 	}
 
-	if (doVersion) {
-		printf("ipcalc %s\n", VERSION);
-		return 0;
-	}
-
 	if (familyIPv6 && familyIPv4) {
```
Comment (nmav) [inline ipcalc.c]: r+

File: netsplit.c [hunk 8]
```diff
@@ -237,7 +237,7 @@ void show_split_networks_v6(unsigned split_prefix, const struct ip_info_st *info
 		count++;
 	}
 
-	dist_printf("\nTotal:  \t", "%u\n", count);
-	dist_printf("Hosts/Net:\t", "%s\n", ipv6_prefix_to_hosts(buf, sizeof(buf), split_prefix));
+	color_printf(ctl, KMAG, "\nTotal:  \t", "%u\n", count);
+	color_printf(ctl, KMAG, "Hosts/Net:\t", "%s\n", ipv6_prefix_to_hosts(buf, sizeof(buf), split_prefix));
 }
```
Comment (nmav) [inline netsplit.c]: r+
I like this change.
A question is why do you bother with such a change for a command line tool? Are you using parts as library?

File: ipcalc.c [hunk 9]
```diff
@@ -573,29 +573,6 @@ char *ipv6_prefix_to_hosts(char *hosts, unsigned hosts_size, unsigned prefix)
 	return hosts;
 }
```
Comment (nmav) [inline ipcalc.c]: I find flags quite easier to handle and manipulate. That is, you can mask multiple easier, and it is harder to forget unused variables around. The whole discussion is bikeshedding though, one way or another is fine. Unless you really have a good argument, I'd keep the existing way for no particular reason.

r-

File: ipcalc-geoip.c [hunk 10]
```diff
@@ -27,6 +27,9 @@
 #include <netinet/in.h>
 #include <arpa/inet.h>
 #include <stdarg.h>
+#include <error.h>
```
Comment (nmav) [inline ipcalc-geoip.c]: These are glibc functions not intended for portable programs.  They are nice, but I don't see a significant improvement to justify losing the portability.

File: ipcalc.c [hunk 11]
```diff
@@ -916,12 +915,8 @@ int get_ipv6_info(struct ipcalc_control *ctl, const char *ipStr,
 #endif
 
 	if (ctl->resolve_host) {
-		info->hostname = get_hostname(AF_INET6, &ip6);
+		info->hostname = get_hostname(ctl, &ip6, ipStr);
 		if (info->hostname == NULL) {
-			if (!ctl->beSilent) {
-				error(0, 0, "cannot find hostname for %s", ipStr);
-				herror(errBuf);
-			}
```
Comment (nmav) [inline ipcalc.c]: r+

File: netsplit.c [hunk 12]
```diff
@@ -216,8 +216,8 @@ void show_split_networks_v6(const struct ipcalc_control *ctl, unsigned split_pre
 		v6add(&start, &sdiff);
 
 		j = 0;
-		for (k = 0; k < 16; k+=2)
-			if (end.s6_addr[k] != 0xff && end.s6_addr[k+1] != 0xff)
+		for (k = 0; k < 16; k += 2)
+			if (end.s6_addr[k] != 0xff && end.s6_addr[k + 1] != 0xff)
 				j = 1;
```
Comment (nmav) [inline netsplit.c]: r+

File: ipcalc.h [hunk 13]
```diff
@@ -61,6 +61,7 @@ int geo_end(struct ipcalc_control *ctl);
 #endif
 
 char __attribute__((warn_unused_result)) *safe_strdup(const char *str);
+int __attribute__((__format__(printf, 2, 3))) safe_asprintf(char **strp, const char *fmt, ...);
 
 char *calc_reverse_dns4(struct in_addr ip, unsigned prefix, struct in_addr net, struct in_addr bcast);
 char *calc_reverse_dns6(struct in6_addr *ip, unsigned prefix);
```
Comment (nmav) [inline ipcalc.h]: r+

**Review by nmav — CHANGES_REQUESTED** [review]
Overall the changes are fine; few patches I don't think should be included. I've marked them as such.

**Comment by nmav** [comment]
Moved repo to https://gitlab.com/ipcalc/ipcalc

**Pull request closed without merging**